alf/algorithms/misc_algorithm.py

```diff
@@ -0,0 +1,203 @@
+# Copyright (c) 2019 Horizon Robotics. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from collections import namedtuple
+from typing import Callable
+
+import gin.tf
+import tensorflow as tf
+
+from tf_agents.networks.network import Network
+import tf_agents.specs.tensor_spec as tensor_spec
+
+from alf.algorithms.algorithm import Algorithm, AlgorithmStep, LossInfo
+from alf.utils.data_buffer import DataBuffer
+from tf_agents.trajectories.time_step import StepType
+from alf.utils.encoding_network import EncodingNetwork
+from alf.utils.common import transpose2
+
+MISCInfo = namedtuple("MISCInfo", ["reward"])
+
+
+@gin.configurable
+class MISCAlgorithm(Algorithm):
+    """Mutual Information-based State Control (MISC)
+    Author: Rui Zhao 
+    Work done during a research internship at Horizon Robotics. 
+    The paper is currently under review in a conference.
+
+    This algorithm generates the intrinsic reward based on the mutual information estimation between the states of interests and the context states.
+
+    See Zhao et al "Self-Supervised State-Control through Intrinsic Mutual Information Rewards"
+    """
+
+    def __init__(self,
+                 batch_size,
+                 observation_spec,
+                 action_spec,
+                 soi_spec,
+                 soc_spec,
+                 split_observation_fn: Callable,
+                 network: Network = None,
+                 mi_r_scale=5000.0,
+                 hidden_size=128,
+                 buffer_size=100,
+                 n_objects=1,
+                 name="MISCAlgorithm"):
+        """Create an MISCAlgorithm.
+        
+        Args:
+            batch_size (int): batch size
+            observation_spec (tf.TensorSpec): observation size
+            action_spec (tf.TensorSpec): action size
+            soi_spec (tf.TensorSpec): state of interest size
+            soc_spec (tf.TensorSpec): state of context size
+            split_observation_fn (Callable): split observation function
```

Comment:
Explain what is the input and the output of this function

Reply:
added